src/Compression/CompressedReadBufferBase.cpp

```diff
@@ -187,6 +187,79 @@ size_t CompressedReadBufferBase::readCompressedData(size_t & size_decompressed,
     return size_compressed_without_checksum + sizeof(Checksum);
 }
 
+size_t CompressedReadBufferBase::readCompressedDataBlockHold(size_t & size_decompressed, size_t & size_compressed_without_checksum)
+{
+    UInt8 header_size = ICompressionCodec::getHeaderSize();
+    if (compressed_in->eof() || (compressed_in->available() < (header_size + sizeof(Checksum))))
```

Comment:
We can't align them with the same as "(compressed_in->eof())"
Because in readCompressedDataHold, we need make sure 2 things as below:
First, make sure the whole header located in 'compressed_in->' buffer.
Second, make sure the whole compressed block located in 'compressed_in->' buffer. Otherwise, abandon header and restore original offset of compressed_in

Instead, i created another function which is implemented by extracting the same part between readCompressedData and readCompressedDataHold. I think it would be easier to understand differences between them.
readHeaderAndGetCodecAndSize

please look into the latest commit for details

Reply:
And for the similar functions:
readHeaderAndGetCodec
readHeaderAndGetCodecAndSize

I also want to merge them but it looks 2 lines in readHeaderAndGetCodec may be important:
```
    ProfileEvents::increment(ProfileEvents::CompressedReadBufferBlocks);
    ProfileEvents::increment(ProfileEvents::CompressedReadBufferBytes, size_decompressed);
```
I'm not sure for that , what's your suggestion?

Reply:
`readHeaderAndGetCodec` and `readHeaderAndGetCodecAndSize` are IMHO sufficiently different to keep them as two methods.

Reply:
Thanks for pulling the common code out into `readHeaderAndGetCodecAndSize`.